nemo/collections/tts/models/talknet.py

```diff
@@ -0,0 +1,226 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import torch
+from hydra.utils import instantiate
+from omegaconf import DictConfig
+from torch import nn
+from torch.nn import functional as F
+
+from nemo.collections.asr.data.audio_to_text import AudioToCharWithDursDataset
+from nemo.collections.tts.helpers.helpers import get_mask_from_lengths
+from nemo.collections.tts.models.base import SpectrogramGenerator
+from nemo.core.classes import ModelPT
+from nemo.core.classes.common import typecheck
+from nemo.core.neural_types.elements import EmbeddedTextType, LengthsType, MelSpectrogramType
+from nemo.core.neural_types.neural_type import NeuralType
+
+
+class TalkNetDursModel(ModelPT):
+    """TalkNet's durations prediction pipeline."""
+
+    def __init__(self, cfg: DictConfig, trainer: 'Trainer' = None):  # noqa
+        super().__init__(cfg=cfg, trainer=trainer)
+
+        cfg = self._cfg
+        self.vocab = AudioToCharWithDursDataset.make_vocab(**cfg.train_ds.dataset.params)
+        self.emb = nn.Embedding(len(self.vocab.labels), cfg.d_char)
+        self.model = instantiate(cfg.model)
+        self.rz = nn.Conv1d(cfg.model.params.jasper[-1].filters, 1, kernel_size=1)
+
+    @property
+    def input_types(self):
+        return {
+            'text': NeuralType(('B', 'T'), EmbeddedTextType()),
+            'text_len': NeuralType(('B',), LengthsType()),
+        }
+
+    @property
+    def output_types(self):
+        return {
+            'durs': NeuralType(('B', 'T'), LengthsType()),
+        }
+
+    @typecheck()
+    def forward(self, *, text, text_len):
+        x, x_len = self.emb(text).transpose(1, 2), text_len
+        y, _ = self.model(audio_signal=x, length=x_len)
+        y = self.rz(y)
+        durs = y.squeeze(1)  # Contracting last 1-dim.
+        return durs
+
+    @staticmethod
+    def _log(durs_true, true_len, durs_pred):
+        loss = F.mse_loss(durs_pred, (durs_true + 1).float().log(), reduction='none')
+        mask = get_mask_from_lengths(true_len)
+        loss *= mask.float()
+        loss = loss.sum() / mask.sum()
+
+        durs_pred = durs_pred.exp() - 1
+        durs_pred[durs_pred < 0.0] = 0.0
+        durs_pred = durs_pred.round().long()
+
+        acc = ((durs_true == durs_pred) * mask).sum().float() / mask.sum()
+        acc = acc * 100
+
+        return dict(loss=loss, acc=acc)
```

Comment:
@okuchaiev @titu1994, this is yet another example why we don't need loss classes. It just doesn't make sense since it's unique to every model. I see no reason why we should block script writers from defining their loss inside their ModelPT class.

Reply:
It depends on the domain - tts is one of the domains that frequently used custom losses. I've found the common losses to be well setup inside the common directory to be used for speech command, VAD and speaker recognition, shared with NLP for their classification tasks. 

On the other hand, TTS models rarely ever share losses amongst themselves or their subclasses. So I can see why it appears like there's no need for such classes. 

In any case, isn't it a programming preference ? Since we follow OOP to quite a degree (as does PT), so we could wrap the above 4 lines into a subclass of MSELoss and call super() just as always. It's reallt upto the person, but inside Nemo there should be consistent development so I would normally ask that this loss be made into a class and then used.

Reply:
I agree with @titu1994.  For consistency, we should pull losses into a separate class.  Even if they are packaged in the same source file or defined as a child class, it would still be weird to have TTS losses implemented differently from ASR losses.

The whole point of neural modules is to split out components when the interface between them can be well defined.  Since there is essentially no overhead now in making the loss function separate, we should do so to be forward-thinking and consistent in our design.

Reply:
My take on this: if loss is generalizable (can be use by some other applications), then let's move it into standalone implementation.
Above losses are ad-hoc TalkNet case logic. There is no need to move them into separate functor/class.